ts/a11y/explorer/Highlighter.ts

```diff
@@ -0,0 +1,485 @@
+
+  /**
+   * The Attribute for marking highlighted nodes.
+   */
+  protected ATTR = 'sre-highlight-' + this.counter.toString();
+
+  /**
+   * The foreground color.
+   */
+  private _foreground: string;
+
+  /**
+   * The background color.
+   */
+  private _background: string;
+
+  /**
+   * The maction name/class for a highlighter.
+   */
+  protected mactionName = '';
+
+  /**
+   * List of currently highlighted nodes and their original background color.
+   */
+  private currentHighlights: Highlight[][] = [];
+
+  /**
+   * Highlights a single node.
+   *
+   * @param node The node to be highlighted.
+   * @returns The old node information.
+   */
+  protected abstract highlightNode(node: HTMLElement): Highlight;
+
+  /**
+   * Unhighlights a single node.
+   *
+   * @param highlight The highlight info for the node to be unhighlighted.
+   */
+  protected abstract unhighlightNode(highlight: Highlight): void;
+
+  /**
+   * @override
+   */
+  public highlight(nodes: HTMLElement[]) {
+    this.currentHighlights.push(
+      nodes.map((node) => {
+        const info = this.highlightNode(node);
+        this.setHighlighted(node);
+        return info;
+      })
+    );
+  }
+
+  /**
+   * @override
+   */
+  public highlightAll(node: HTMLElement) {
+    const mactions = this.getMactionNodes(node);
+    for (let i = 0, maction; (maction = mactions[i]); i++) {
+      this.highlight([maction]);
+    }
+  }
+
+  /**
+   * @override
+   */
+  public unhighlight() {
+    const nodes = this.currentHighlights.pop();
+    if (!nodes) {
+      return;
+    }
+    nodes.forEach((highlight: Highlight) => {
+      if (this.isHighlighted(highlight.node)) {
+        this.unhighlightNode(highlight);
+        this.unsetHighlighted(highlight.node);
+      }
+    });
+  }
+
+  /**
+   * @override
+   */
+  public unhighlightAll() {
+    while (this.currentHighlights.length > 0) {
+      this.unhighlight();
+    }
+  }
+
+  /**
+   * @override
+   */
+  public setColor(background: NamedColor, foreground: NamedColor) {
+    this._foreground = getColorString(foreground, DEFAULT_FOREGROUND);
+    this._background = getColorString(background, DEFAULT_BACKGROUND);
+  }
+
+  /**
+   * @override
+   */
+  public get foreground(): string {
+    return this._foreground;
+  }
+
+  /**
+   * @override
+   */
+  public get background(): string {
+    return this._background;
+  }
+
+  /**
+   * Returns the maction sub nodes of a given node.
+   *
+   * @param {HTMLElement} node The root node.
+   * @returns {HTMLElement[]} The list of maction sub nodes.
+   */
+  public getMactionNodes(node: HTMLElement): HTMLElement[] {
+    return Array.from(
+      node.getElementsByClassName(this.mactionName)
+    ) as HTMLElement[];
+  }
+
+  /**
+   * @override
+   */
+  public isMactionNode(node: Element): boolean {
+    const className = node.className || node.getAttribute('class');
+    return className ? !!className.match(new RegExp(this.mactionName)) : false;
```

Comment:
Note that MathJax v4 does not use the class name to identify maction nodes (that was a v2 approach).  In version 4, you could use the tag name being `'mjx-' + this.mactionName` or having an attribute `data-mjx-node` be `this.mactionName`.  Because of this, both the CHTML and SVG highlighters have to override this.   It might be better to either make this an abstract method that the subclass must implement, or make it check for the two approaches that are used.

Reply:
That's the old default from SRE, which worked on a number of the 2.7 renderers. I thought I had caught them all and made them abstract.